### DISCUSSION TITLE
Q: Best way to approach manipulation of coverage instrumentation

### DISCUSSION THREAD

**Discussion by liona24** [AFLplusplus/AFLplusplus#1638] [Q&A]
Hey there, just started checking out the project more in depth.

What I want to do is start hacking on the coverage instrumentation code (mainly for research, trying things out, etc.).
If I were to do this in plain LLVM I would look at the BB instrumentation [here](https://github.com/llvm/llvm-project/blob/main/llvm/lib/Transforms/Instrumentation/SanitizerCoverage.cpp#L990) and additionally modify the callbacks (`__sanitizer_*` in [FuzzerTracePC](https://github.com/llvm/llvm-project/blob/main/compiler-rt/lib/fuzzer/FuzzerTracePC.cpp)). This would give me a modified fuzzer in no time.

What is the best approach here for AFL? I feel a little bit overwhelmed and figured fastest way to get going would be just asking.

There seem to be many places where coverage seems to be ejected but I am not sure what needs to be modified (i.e. minimal set)

**Comment by vanhauser-thc** [comment]
similar.
the sanitizer callback functions are in instrumentation/afl-compiler.rt.o.c and the main coverage instrumentation is in instrumentation/SanitizerCoveragePCGUARD.so.cc which is based on the SanitizerCoverage.cc llvm but being improved on.
